Source/WebCore/accessibility/AccessibilityNodeObject.cpp

```diff
@@ -1819,6 +1820,24 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
     if (isTreeItem() && ariaLabel.isEmpty() && ariaLabeledByAttribute().isEmpty())
         textOrder.append(AccessibilityText(accessibleNameForNode(*node), AccessibilityTextSource::Alternative));
 
+    // Dialog, alert dialog and article elements may derive their name from the first descendant node
+    // that is a heading (depth-first search, preorder traversal).
+    if (AccessibilityObject::accessibleNameDerivesFromHeading()) {
+        if (auto* rootNode = dynamicDowncast<ContainerNode>(node)) {
```

Comment:
Maybe `containerNode` instead of `rootNode`? "Root" has strong meanings about the top-level object in the tree, which I don't think is the intention here.